### PR TITLE
Partialy fix issue #23334 - isort pandas/core/dtypes directory

### PR DESCRIPTION
**Pull request by alexander-ponomaroff**
The imports have been sorted with isort in the pandas/core/dtypes directory.

Moving onto sorting the pandas/core/groupby directory.

- [x] partly #23334
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by alimcmaster1** [comment]
Thanks @alexander-ponomaroff ! Lets wait for CI to run then i'm happy to review this.

**Comment by alexander-ponomaroff** [comment]
Thank you @alimcmaster1 . Could you please explain what "passes git diff upstream/master -u -- "*.py" | flake8 --diff" means? I noticed that you have this checked in your pull requests.

**Comment by alexander-ponomaroff** [comment]
> Hey - please see the "Python (PEP8)" section in the [contributing guide](https://pandas.pydata.org/pandas-docs/stable/contributing.html) and further info on [flake8 here](http://flake8.pycqa.org/en/latest/) - hope that helps.

Thanks, just performed the check locally. But I'm assuming the pep8speaks bot, which is the second comment of this pr does the same thing here?

**Comment by alimcmaster1** [comment]
https://travis-ci.org/pandas-dev/pandas/pull_requests , you can look at the CI run off the back of previous PRs to get an idea of how long it will take, feel free to wait till tomorrow there is no rush

**Comment by jreback** [comment]
pls merge master and ping on green.

**Comment by alimcmaster1** [comment]
`from pandas.core.dtypes.generic import ABCPeriodArray` in common.py it looks like you have removed this import? Did isort do this?

**Comment by alexander-ponomaroff** [comment]
@alimcmaster1 Not sure what happened, I did not remove it on purpose.

**Comment by alexander-ponomaroff** [comment]
@alimcmaster1 I just added ABCPeriodArray back and fixed another conflict.

**Comment by thoo** [comment]
@alexander-ponomaroff  You might want to rerun `isort` to fix travis CI. I did on some of my PRs and they are passed now. It might be due to `upstream/merge`.

**Comment by alexander-ponomaroff** [comment]
@thoo I reran isort, but the CI is still failing. @alimcmaster1 Do you have any suggestions on what I should do?

**Comment by thoo** [comment]
You can go to the log file and see what is go wrong. For example, this is the first one failed. 

```
[31m[1m_____________ TestDataFrameCombineFirst.test_combine_first_period ______________[0m
[gw1] linux -- Python 3.7.1 /home/travis/miniconda3/envs/pandas/bin/python

self = <pandas.tests.frame.test_combine_concat.TestDataFrameCombineFirst object at 0x7f06cdf74470>

[1m    def test_combine_first_period(self):[0m
[1m        data1 = pd.PeriodIndex(['2011-01', 'NaT', '2011-03',[0m
[1m                                '2011-04'], freq='M')[0m
[1m        df1 = pd.DataFrame({'P': data1}, index=[1, 3, 5, 7])[0m
[1m        data2 = pd.PeriodIndex(['2012-01-01', '2012-02',[0m
[1m                                '2012-03'], freq='M')[0m
[1m        df2 = pd.DataFrame({'P': data2}, index=[2, 4, 5])[0m
[1m    [0m
[1m        res = df1.combine_first(df2)[0m
[1m        exp_dts = pd.PeriodIndex(['2011-01', '2012-01', 'NaT',[0m
[1m                                  '2012-02', '2011-03', '2011-04'],[0m
[1m                                 freq='M')[0m
[1m        exp = pd.DataFrame({'P': exp_dts}, index=[1, 2, 3, 4, 5, 7])[0m
[1m>       tm.assert_frame_equal(res, exp)[0m
[1m[31mE       AssertionError: Attributes are different[0m
[1m[31mE       [0m
[1m[31mE       Attribute "dtype" are different[0m
[1m[31mE       [left]:  int64[0m
[1m[31mE       [right]: period[M][0m

[1m[31mpandas/tests/frame/test_combine_concat.py[0m:775: AssertionError
```

**Comment by thoo** [comment]
I might rebase :
```
git fetch upstream
git merge upstream/master
```
and after that rerun `isort` to make sure everything is as you wanted. It should not be but it won't hurt.

**Comment by alexander-ponomaroff** [comment]
@jreback Hello Jeff, I cannot seem to identify the cause of CI failing, it recently started failing after I resolved the second conflict with the master branch.

**Comment by alexander-ponomaroff** [comment]
I notice that this was changed a few days ago, and it's now failing here:
https://github.com/pandas-dev/pandas/blame/5d84bc08ea7c45b0d2b3aa021a425b11531336f1/pandas/tests/frame/test_combine_concat.py#L776

@TomAugspurger you worked on this, do you have any advice for me? Or know why this failure was triggered in my pr?

**Comment by alexander-ponomaroff** [comment]
@alimcmaster1 I reran everything from scratch and the previously triggered error doesn't trigger anymore. However there is an error now `KeyError: <class 'pandas._libs.tslibs.timestamps.Timestamp'>`, not sure what this has to do with anything I did. Do you have any suggestions?

**Comment by TomAugspurger** [comment]
You may need to rebuild the C extensions: `python setup.py build_ext
--inplace -j 4`

**Comment by alexander-ponomaroff** [comment]
@TomAugspurger I get this error: `pandas/_libs/window.cpp:612:10: fatal error: 'ios' file not found`

**Comment by TomAugspurger** [comment]
Others have reported that. What version of MacOS? You may need to update
your command line tools.

**Comment by alexander-ponomaroff** [comment]
@TomAugspurger High Sierra 10.13.6

**Comment by TomAugspurger** [comment]
What does `xcode-select --install` output?

**Comment by alexander-ponomaroff** [comment]
@TomAugspurger Got it, I updated the command line tools and it worked. I rebuilt the C extensions. Where do I go from here?

**Comment by alexander-ponomaroff** [comment]
@TomAugspurger Btw, the error `KeyError: <class 'pandas._libs.tslibs.timestamps.Timestamp'>` is in one of the tests from Azure Pipelines, in case you thought that it was local. I just noticed that you're replying from an email. That's the only check I fail from the CI checks.

File: pandas/core/dtypes/concat.py [hunk 1]
```diff
@@ -1,27 +1,21 @@
 """
-Utility functions related to concat
-"""
+    Utility functions related to concat
+    """
```
Comment (alimcmaster1) [inline pandas/core/dtypes/concat.py]: Don't think this indenting is required?

Comment (alexander-ponomaroff) [reply pandas/core/dtypes/concat.py]: Will fix, accidental.

**Comment by alimcmaster1** [comment]
"Btw, the error KeyError: <class 'pandas._libs.tslibs.timestamps.Timestamp'> is in one of the tests from Azure Pipelines"

I don't think the pipeline is failing because of this. [Here](https://github.com/pandas-dev/pandas/pull/23331/checks?check_run_id=27234672) is an example of a pipeline with the same test errors but the pipeline succeeds.

Your [pipeline](https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=2073&view=logs) seems to be failing "WindowsPy27 py36_np121" builds with the error:

2018-10-29T15:30:51.4397165Z Could not find conda environment: pandas
2018-10-29T15:30:51.4398019Z You can list all discoverable environments with `conda info --envs`.
2018-10-29T15:30:56.1307704Z ModuleNotFoundError: No module named 'numpy'

Maybe @TomAugspurger could restart that for you? As i'm really not sure what the issue is there.

**Comment by alexander-ponomaroff** [comment]
@alimcmaster1 Wow, finally! Everything passed. Thanks for you help.

**Comment by alexander-ponomaroff** [comment]
@jreback Everything passed now, could you please merge if everything is good?

File: pandas/core/dtypes/missing.py [hunk 2]
```diff
@@ -1,28 +1,20 @@
 """
-missing types & inference
-"""
```
Comment (jreback) [inline pandas/core/dtypes/missing.py]: why did this change?

File: pandas/core/dtypes/api.py [hunk 3]
```diff
@@ -1,59 +1,15 @@
...
-                     is_re,
-                     is_re_compilable,
-                     is_dict_like,
-                     is_iterator,
-                     is_file_like,
-                     is_array_like,
-                     is_list_like,
-                     is_hashable,
-                     is_named_tuple)
+# categorical; interval; datetimelike; string-like; sparse; numeric types; like
```
Comment (jreback) [inline pandas/core/dtypes/api.py]: i am not sure of the value of this comment

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/dtypes/missing.py (hunk 2 above)
Comment (alexander-ponomaroff) [reply pandas/core/dtypes/missing.py]: Accidental, will fix. When I had errors, I copy pasted imports from master and reran isort to attempt to fix the errors and the spacing on some of the comments messed up.

File: pandas/core/dtypes/api.py (hunk 3 above)
Comment (alexander-ponomaroff) [reply pandas/core/dtypes/api.py]: This comment was generated by isort, will delete.

**Comment by alexander-ponomaroff** [comment]
@jreback The requested changes have been changed, all checks pass.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @alexander-ponomaroff

**Pull request merged**